### PR TITLE
feat: printing all the errors from goparser

### PR DESCRIPTION
**Pull request by Villaquiranm**
This PR aims to print all the errors that are found by the go/parser package.
Currently the `parser.ParseFile` function returns an `scanner.ErrorList` object that only prints the first error and a count of the remaining errors. #1933 everything is better explained here.

**For the changes this PR introduces**:
- Create **issueWithError** function.  This function will wrap the lines of code that processed an error (before this PR changes) in order to convert it on an **lintIssue** structure. This existent code applies a regex to localize the parse error more precisely.
- The PR changes the behavior of run, test, and lint commands. Not they exit with a non-zero status code as soon as we found some error while parsing the files. This in order to not having the whole stacktrace for some parsing file error, improving visibility.
- Add a new case on `gnovm/cmd/gno/lint.go.catchRuntimeError` to handle and print all errors of an error with type `scanner.ErrorList`
- remove some wrapping during panic calls. This was done by simplicity some of the calls were doing `panic(errors.Wrap(err, "parsing file "+mfile.Name))` here err has the type scanner.ErrorList but as we're wrapping it it would be more complex and difficult to read on **catchRuntimeError** to identify an error with the type ErrorList

**Results:** (Using issue example)
- lint:
```cmd
-- NOW
./.: missing 'gno.mod' file (code=1).
test.gno:6: expected ';', found example (code=2).
test.gno:7: expected '}', found 'EOF' (code=2).
exit status 1

---- BEFORE
./.: missing 'gno.mod' file (code=1).
test.gno:6: expected ';', found example (and 1 more errors) (code=2).
```
- run:
```cmd
-- NOW
test.gno:6: expected ';', found example (code=2).
test.gno:7: expected '}', found 'EOF' (code=2).
exit status 1

---- BEFORE
panic: test.gno:6:5: expected ';', found example (and 1 more errors)

goroutine 1 [running]:
github.com/gnolang/gno/gnovm/pkg/gnolang.MustReadFile(...)
        /Users/miguel/gnorigin/gnovm/pkg/gnolang/go2gno.go:49
main.parseFiles({0x140003b1720, 0x1, 0x1400031fb48?})
        /Users/miguel/gnorigin/gnovm/cmd/gno/run.go:132 +0x27c
main.parseFiles({0x140003b09a0, 0x1, 0x0?})
        /Users/miguel/gnorigin/gnovm/cmd/gno/run.go:121 +0x158
main.execRun(0x14000373230, {0x140003b09a0, 0x1, 0x1}, {0x100ca3e68, 0x140003b3bd0})
        /Users/miguel/gnorigin/gnovm/cmd/gno/run.go:89 +0x174
main.newRunCmd.func1({0x0?, 0x1400019c140?}, {0x140003b09a0?, 0x0?, 0x0?})
        /Users/miguel/gnorigin/gnovm/cmd/gno/run.go:35 +0x3c
github.com/gnolang/gno/tm2/pkg/commands.(*Command).Run(0x1400031fdc8?, {0x100c9af18?, 0x101249700?})
        /Users/miguel/gnorigin/tm2/pkg/commands/command.go:247 +0x17c
github.com/gnolang/gno/tm2/pkg/commands.(*Command).Run(0x140003ae2c0?, {0x100c9af18?, 0x101249700?})
        /Users/miguel/gnorigin/tm2/pkg/commands/command.go:251 +0x12c
github.com/gnolang/gno/tm2/pkg/commands.(*Command).ParseAndRun(0x140003ae6e0?, {0x100c9af18, 0x101249700}, {0x1400019c130?, 0x140003ae790?, 0x140003ae840?})
        /Users/miguel/gnorigin/tm2/pkg/commands/command.go:132 +0x4c
github.com/gnolang/gno/tm2/pkg/commands.(*Command).Execute(0x100ca3e68?, {0x100c9af18?, 0x101249700?}, {0x1400019c130?, 0x1011a5ef8?, 0x140000021a0?})
        /Users/miguel/gnorigin/tm2/pkg/commands/command.go:114 +0x28
main.main()
        /Users/miguel/gnorigin/gnovm/cmd/gno/main.go:13 +0x6c
```
<details><summary>Contributors' checklist...</summary>

- [ ] Added new tests, or not needed, or not feasible
- [ ] Provided an example (e.g. screenshot) to aid review or the PR is self-explanatory
- [ ] Updated the official documentation or not needed
- [ ] No breaking changes were made, or a `BREAKING CHANGE: xxx` message was included in the description
- [ ] Added references to related issues and PRs
- [ ] Provided any useful hints for running manual tests
- [ ] Added new benchmarks to [generated graphs](https://gnoland.github.io/benchmarks), if any. More info [here](https://github.com/gnolang/gno/blob/master/.benchmarks/README.md).
</details>

### CODE REVIEW

**Comment by moul** [comment]
Great addition! It's definitely needed. Could you simplify the implementation?

File: gnovm/cmd/gno/util.go [hunk 1]
```diff
@@ -326,3 +326,44 @@ func prettySize(nb int64) string {
 	}
 	return fmt.Sprintf("%.1f%c", float64(nb)/float64(div), "kMGTPE"[exp])
 }
+
+type issueAdder struct {
```
[the remark this answers is not in the document]
Comment (Villaquiranm) [reply gnovm/cmd/gno/util.go:330]: Thanks a lot for your quick review :)

I think the code misleads a little bit, but we already have an error list (Similar to go.uber.org/multierr) coming from the go parser (**scanner.ErrorList**). 
What issuerAdder does actually is simply log the error to the stderr. maybe we should rename it as issueLogger or as you suggest make something simpler.

What I was aiming was to refactor these two pieces of code
```go
addIssue := func(issue lintIssue) {
     hasError = true
     fmt.Fprint(io.Err(), issue.String()+"\n")
}
```
and 
```go
var issue lintIssue
issue.Confidence = 1
issue.Code = lintGnoError

parsedError := strings.TrimSpace(err.Error())
parsedError = strings.TrimPrefix(parsedError, pkgPath+"/")

matches := reParseRecover.FindStringSubmatch(parsedError)
if len(matches) == 4 {
	sourcepath := guessSourcePath(pkgPath, matches[1])
	issue.Location = fmt.Sprintf("%s:%s", sourcepath, matches[2])
	issue.Msg = strings.TrimSpace(matches[3])
} else {
	issue.Location = fmt.Sprintf("%s:0", filepath.Clean(pkgPath))
	issue.Msg = err.Error()
}

addIssue(issue)
```
That were used on lint cmd and reuse them easily on run and test cmds

[the remark this answers is not in the document]
Comment (moul) [reply gnovm/cmd/gno/util.go:330]: I believe the intermediary struct should be eliminated. The parsing makes sense, but it should append to a multierr or write to stderr directly. Consider simplifying the code by removing the intermediary struct entirely. Aim to add as few new functions as possible, ideally none or just one, focused on aiding parsing.

I'm unsure if my proposal is feasible, but the intermediary structure seems problematic to me.

[the remark this answers is not in the document]
Comment (Villaquiranm) [reply gnovm/cmd/gno/util.go]: Hello thanks again for the review, I removed all the structs,
Just Added a new constructor  **issueWithError** (Takes and error and a path as parameter to create an struct issue, as before) using this existent code:
```go
var issue lintIssue
issue.Confidence = 1
issue.Code = lintGnoError

parsedError := strings.TrimSpace(err.Error())
parsedError = strings.TrimPrefix(parsedError, pkgPath+"/")

matches := reParseRecover.FindStringSubmatch(parsedError)
if len(matches) == 4 {
	sourcepath := guessSourcePath(pkgPath, matches[1])
	issue.Location = fmt.Sprintf("%s:%s", sourcepath, matches[2])
	issue.Msg = strings.TrimSpace(matches[3])
} else {
	issue.Location = fmt.Sprintf("%s:0", filepath.Clean(pkgPath))
	issue.Msg = err.Error()
}
```

File: gnovm/tests/integ/several-lint-errors/main.gno [hunk 2]
```diff
@@ -0,0 +1,7 @@
+// a.gno example
```
Comment (thehowl) [inline gnovm/tests/integ/several-lint-errors/main.gno:1]:
```suggestion
```

**Comment by thehowl** [comment]
Thank you for your contribution!

**Review by petar-dambovaliev — APPROVED** [review]

File: gnovm/tests/integ/several-lint-errors/main.gno (hunk 2 above)
Comment (Villaquiranm) [reply gnovm/tests/integ/several-lint-errors/main.gno]: fixed on [91a66af](https://github.com/gnolang/gno/pull/2011/commits/91a66afa7749be834a603cd69f64ca6f897e2918)

File: gnovm/cmd/gno/lint.go [hunk 3]
```diff
@@ -229,3 +215,23 @@ func (i lintIssue) String() string {
 	// TODO: consider crafting a doc URL based on Code.
 	return fmt.Sprintf("%s: %s (code=%d).", i.Location, i.Msg, i.Code)
 }
+
+func issueWithError(pkgPath string, err error) lintIssue {
```
[the remark this answers is not in the document]
Comment (Villaquiranm) [reply gnovm/cmd/gno/lint.go]: fixed on [91a66af](https://github.com/gnolang/gno/pull/2011/commits/91a66afa7749be834a603cd69f64ca6f897e2918)

File: gnovm/cmd/gno/lint.go [hunk 4]
```diff
@@ -81,16 +79,18 @@ func execLint(cfg *lintCfg, args []string, io commands.IO) error {
...
 				Location:   pkgPath,
 				Msg:        "missing 'gno.mod' file",
-			})
+			}
+			fmt.Fprint(io.Err(), issue.String()+"\n")
 		}
 
 		// Handle runtime errors
-		catchRuntimeError(pkgPath, addIssue, func() {
+		hasError = hasError || catchRuntimeError(pkgPath, io.Err(), func() {
```
[the remark this answers is not in the document]
Comment (Villaquiranm) [reply gnovm/cmd/gno/lint.go]: fixed on [91a66af](https://github.com/gnolang/gno/pull/2011/commits/91a66afa7749be834a603cd69f64ca6f897e2918)

By the way great catch :)

**Comment by Villaquiranm** [comment]
Just a little remark on this PR,
With @omarsy we realized that even with this PR we could have some linter errors that are now shown.

This happens when having errors on different files
Let's say: **a.gno** (2 lint errors) **b.gno** (1 lint) error. We will only show all lint errors of only the first file.

I would be glad to work on this again in order to have all errors on all folder files something more Go like

**Review by thehowl — APPROVED** [review]

**Comment by thehowl** [comment]
> We will only show all lint errors of only the first file.

Feel free to make a follow up :)

**Pull request merged**